buchschloss/core.py

```diff
@@ -1097,6 +1099,218 @@ def view_str(member, *, login_context):
+        script_lc = LoginType.SCRIPT(
+            script_lc_level, name=script.name, invoker=login_context)
+        ui_callbacks = callbacks or cls.callbacks
+        get_name_prefix = 'script-data::{}::'.format(script.name)
+        script_config = config.scripts.lua.get(script.name).mapping
+        if ScriptPermissions.STORE in script.permissions:
+            edit_func = partial(Script.edit, script.name, login_context=internal_priv_lc)
+            add_storage = (
+                lambda: Script.view_ns(script.name, login_context=internal_priv_lc).storage,
+                lambda data: edit_func(storage=data),
+            )
+        else:
+            add_storage = None
+        ns = lua.execute_script(
+            script.code,
+            script_lc,
+            add_ui=(None if ui_callbacks is None else (ui_callbacks, get_name_prefix)),
+            add_storage=add_storage,
+            add_requests=(ScriptPermissions.REQUESTS in script.permissions),
+            add_config=script_config,
+        )
+        if function is not None:
+            try:
+                ns[function]()
+            except Exception as e:
+                if config.debug:
+                    raise
+                logging.error('error executing script function: ' + str(e))
+                display = ':'.join((script.name, function))
+                raise BuchSchlossError('Script::execute', 'script_{}_exec_problem', display)
+
+
+class DataNamespace:
+    """class for data namespaces returned by view_ns"""
+    def __init__(self, ans: T.Type[ActionNamespace],
+                 raw_data: T.Any,
+                 login_context: LoginContext):
+        """initialize this namespace with data from the database"""
+        self._data = raw_data
+        self._handlers = self.data_handling[ans]
+        self._login_context = login_context
+
+    def __eq__(self, other):
+        if isinstance(other, DataNamespace):
+            return self._data == other._data
+        elif isinstance(other, type(self.id)):
+            return self.id == other
+        else:
+            return NotImplemented
+
+    def __dir__(self) -> T.Iterable[str]:
+        return set(itertools.chain(
+            super().__dir__(),
+            ['id'],  # some already have it -> set
+            *self._handlers.values(),
+        ))
+
+    def __hash__(self):
+        return hash(self._data)
+
+    def __str__(self):
+        return str(self._data)
+
+    def __getattr__(self, item):
+        def get_id(obj):
+            """I hate myself for not calling the IDs 'id'"""
+            return getattr(obj, type(obj).pk_name)
+
+        if item in self._handlers['allow']:
+            return getattr(self._data, item)
+        elif item in self._handlers.get('wrap_iter', {}):
+            view_ns = self._handlers['wrap_iter'][item].view_ns
+            return [view_ns(get_id(o), login_context=self._login_context)
+                    for o in getattr(self._data, item)]
```

Comment:
Is this efficient / do all those ``o``s *already* have the Data I'm getting from the DB a second time with ``view_ns``?